contracts/libraries/LibCraft.sol

```diff
@@ -44,16 +44,19 @@ library LibCraft {
         uint256[] rewardAmounts;
     }
 
+    function hashAssetId(AssetData memory _asset) internal pure returns(uint256) {
+        return uint256(keccak256(abi.encodePacked(_asset.content, _asset.tokenId)));
+    }
+
     function verifySalvageableAsset(SalvageableAsset memory _asset) internal pure {
         // No need to check the validity of the contract. All registered contracts are Content contracts. If we get
-        // here, it means we've verified the asset and reward assets correctly.
-        require(_asset.salvageType < uint256(SalvageType.Max), "Invalid Salvage Type");
-        require(_asset.rewards.length > 0, "Invalid rewards length");
-        for (uint256 i = 0; i < _asset.rewards.length; ++i) {
-            require(_asset.rewards[i].probability > 0 && _asset.rewards[i].probability <= 1000000, "Invalid probability");
-            require(_asset.rewards[i].amount > 0, "Invalid reward amount");
-            require(_asset.rewards[i].asset.content != address(0), "Invalid content address");
-            require(_asset.rewards[i].asset.tokenId != 0, "Invalid token id");
+        // here, it means we've verified the asset and output assets correctly.
+        require(_asset.salvageType < uint256(SalvageType.Max), "Error: Invalid Salvage Type");
+        for (uint256 i = 0; i < _asset.outputs.length; ++i) {
+            require(_asset.outputs[i].probability > 0 && _asset.outputs[i].probability <= 1e6, "Error: Invalid probability");
+            require(_asset.outputs[i].amount > 0, "Error: Invalid reward amount");
+            require(_asset.outputs[i].asset.content != address(0), "Error: Invalid content address");
+            require(_asset.outputs[i].asset.tokenId != 0, "Error: Invalid token id");
```

Comment:
I might still just not understand this very well but if it's the salveageable asset's content address and tokenId that gets hashed, shouldn't we be checking whether _asset.asset.tokenId != 0 and not the output item's tokenId?

Reply:
Here, we're just checking if the asset exists. However, token id of 0 is totally possible so this is a bug on my part. I will delete it.